selfdrive/car/nissan/interface.py

```diff
@@ -0,0 +1,193 @@
+
+    return ret
+
+  # returns a car.CarState
+  def update(self, c, can_strings):
+    self.pt_cp.update_strings(can_strings)
+    self.cam_cp.update_strings(can_strings)
+    self.adas_cp.update_strings(can_strings)
+
+    self.CS.update(self.pt_cp, self.adas_cp, self.cam_cp)
+
+    # create message
+    ret = car.CarState.new_message()
+
+    ret.canValid = self.pt_cp.can_valid and self.adas_cp.can_valid and self.cam_cp.can_valid
+
+    # speeds
+    ret.vEgo = self.CS.v_ego
+    ret.aEgo = self.CS.a_ego
+    ret.vEgoRaw = self.CS.v_ego_raw
+    ret.yawRate = self.VM.yaw_rate(self.CS.angle_steers * CV.DEG_TO_RAD, self.CS.v_ego)
+    ret.standstill = self.CS.standstill
+    ret.wheelSpeeds.fl = self.CS.v_wheel_fl
+    ret.wheelSpeeds.fr = self.CS.v_wheel_fr
+    ret.wheelSpeeds.rl = self.CS.v_wheel_rl
+    ret.wheelSpeeds.rr = self.CS.v_wheel_rr
+
+    # steering wheel
+    ret.steeringAngle = self.CS.angle_steers
+
+    # torque and user override. Driver awareness
+    # timer resets when the user uses the steering wheel.
+    ret.steeringPressed = self.CS.steer_override
+    ret.steeringTorque = self.CS.steer_torque_driver
+
+    ret.gas = self.CS.pedal_gas
+    ret.gasPressed = self.CS.user_gas_pressed
+
+    # cruise state
+    ret.cruiseState.enabled = bool(self.CS.acc_active)
+    ret.cruiseState.speed = self.CS.v_cruise_pcm * CV.KPH_TO_MS
+    ret.cruiseState.available = bool(self.CS.main_on)
+    ret.cruiseState.speedOffset = 0.
+
+    ret.leftBlinker = self.CS.left_blinker_on
+    ret.rightBlinker = self.CS.right_blinker_on
+    ret.seatbeltUnlatched = self.CS.seatbelt_unlatched
+    ret.doorOpen = self.CS.door_open
+
+    buttonEvents = []
+
+    # blinkers
+    if self.CS.left_blinker_on != self.CS.prev_left_blinker_on:
+      be = car.CarState.ButtonEvent.new_message()
+      be.type = ButtonType.leftBlinker
+      be.pressed = self.CS.left_blinker_on
+      buttonEvents.append(be)
+
+    if self.CS.right_blinker_on != self.CS.prev_right_blinker_on:
+      be = car.CarState.ButtonEvent.new_message()
+      be.type = ButtonType.rightBlinker
+      be.pressed = self.CS.right_blinker_on
+      buttonEvents.append(be)
+
+    be = car.CarState.ButtonEvent.new_message()
+    be.type = ButtonType.accelCruise
+    buttonEvents.append(be)
+
+
+    events = []
+    if ret.seatbeltUnlatched:
+      events.append(create_event('seatbeltNotLatched', [ET.NO_ENTRY, ET.SOFT_DISABLE]))
+
+    if ret.doorOpen:
+      events.append(create_event('doorOpen', [ET.NO_ENTRY, ET.SOFT_DISABLE]))
+
+    if self.CS.acc_active and not self.acc_active_prev:
+      events.append(create_event('pcmEnable', [ET.ENABLE]))
+    if not self.CS.acc_active:
+      events.append(create_event('pcmDisable', [ET.USER_DISABLE]))
+
+    # disable on gas pedal rising edge
+    if (ret.gasPressed and not self.gas_pressed_prev):
+      events.append(create_event('pedalPressed', [ET.NO_ENTRY, ET.USER_DISABLE]))
+
```

Comment:
can we also disable on brake pressed when speed > 0 (see Honda)? it's a change that is coming for all cars and it'll be consistent with safety.